eras/babbage/impl/testlib/Test/Cardano/Ledger/Babbage/Imp/UtxowSpec/Valid.hs

```diff
@@ -26,44 +48,159 @@ spec ::
   ) =>
   SpecWith (ImpInit (LedgerSpec era))
 spec = describe "Valid" $ do
-  it "Native reference scripts must be witnessed" $ do
+  it "Native reference scripts must not be witnessed" $ do
     addr <- freshKeyAddr_
     let
-      timelock = fromNativeScript @era $ RequireAnyOf []
-      txOut =
-        mkCoinTxOut addr (inject $ Coin 15_000_000)
-          & referenceScriptTxOutL .~ SJust timelock
-    tx0 <-
+      script = nativeFails -- Tx will fail if this is witnessed
+      txOutRef =
+        mkBasicTxOut addr mempty
+          & referenceScriptTxOutL .~ SJust script
+    txInitial <-
       submitTx $
         mkBasicTx mkBasicTxBody
-          & bodyTxL . outputsTxBodyL .~ [txOut]
+          & bodyTxL . outputsTxBodyL .~ [txOutRef]
+    submitTx_ $
+      mkBasicTx mkBasicTxBody
+        & bodyTxL . referenceInputsTxBodyL .~ [txInAt 0 txInitial]
+
+  it "Spend native script output with reference script" $ do
+    addr <- freshKeyAddr_
     let
-      txIn = mkTxInPartial (txIdTx tx0) 0
-      tx1 =
+      script = nativeSucceeds
+      txOut =
+        mkBasicTxOut (mkAddr (hashScript script) StakeRefNull) mempty
+      txOutRef =
+        mkBasicTxOut addr mempty
+          & referenceScriptTxOutL .~ SJust script
+    txInitial <-
+      submitTx $
         mkBasicTx mkBasicTxBody
-          & bodyTxL . referenceInputsTxBodyL .~ [txIn]
-    submitTx_ tx1
-
-  it "Inline datum" $ do
-    const $ pendingWith "not implemented yet"
-
-  it "Reference script" $ do
-    const $ pendingWith "not implemented yet"
-
-  it "Inline datum and ref script" $ do
-    const $ pendingWith "not implemented yet"
+          & bodyTxL . outputsTxBodyL .~ [txOut, txOutRef]
+    submitTx_ $
+      mkBasicTx $
+        mkBasicTxBody
+          & inputsTxBodyL .~ [txInAt 0 txInitial]
+          & referenceInputsTxBodyL .~ [txInAt 1 txInitial]
 
   it "Reference input with data hash, no data witness" $ do
-    const $ pendingWith "not implemented yet"
+    addr <- freshKeyAddr_
+    let
+      datumValue = Data @era $ PV1.B "abcde"
+      datumHash = hashData datumValue
+      txOut =
+        mkBasicTxOut addr mempty
+      txOutDatum =
+        mkBasicTxOut addr mempty
+          & datumTxOutL .~ DatumHash datumHash
+    txInitial <-
+      submitTx $
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . outputsTxBodyL .~ [txOut, txOutDatum]
+    tx <-
+      submitTx $
+        mkBasicTx mkBasicTxBody
+          & bodyTxL . inputsTxBodyL .~ [txInAt 0 txInitial]
+          & bodyTxL . referenceInputsTxBodyL .~ [txInAt 1 txInitial]
+    -- Verify that a witness wasn't added by fixup
+    tx ^. witsTxL . datsTxWitsL . unTxDatsL . at datumHash `shouldBe` Nothing
```

Comment:
Don't you think using `lookup` makes it a bit more clear what is going on? Also we have a helper for the `Nothing` case
```suggestion
    expectNothing (Map.lookup datumHash (tx ^. witsTxL . datsTxWitsL . unTxDatsL)
```
Feel free to ignore this suggestion, just making an opinionated observation :wink:

Reply:
I left it this way to highlight the parallelism with the immediately following test which has a `Just` value, and which has a line that sets the datum using a compound lens and then a line that checks it using the same lens. I think the all-lens approach has advantages for setting, ie

```haskell
  & witsTxL . datsTxWitsL . unTxDatsL . at datumHash .~ Just datumValue
```

is more readable than

```haskell
  & witsTxL . datsTxWitsL . unTxDatsL %~ Map.insert datumHash (Just datumValue)
```